ocrd/ocrd/workspace.py

```diff
@@ -851,7 +852,15 @@ def save_image_file(self, image,
         if not force and self.overwrite_mode:
             force = True
         image_bytes = io.BytesIO()
-        image.save(image_bytes, format=MIME_TO_PIL[mimetype])
+        if dpi and not isinstance(dpi, tuple):
+            dpi = (dpi, dpi)
+        elif not dpi:
+            # TODO - brittle, will this always find the original image?
+            orig_img_file = self.mets.find_files(pageId=page_id, mimetype='//^image')[0]
+            orig_img_pil = self.resolve_image_exif(orig_img_file.url)
```

Comment:
You're right. I did not think right. There might be many images (even derived images) for that pageId already. It's probably better than nothing though. But we should allow empty results here, or at least raise an appropriate error.

Also, I wonder whether this does not make preprocessing much more inefficient. I don't know about lxml's `xpath` and `iter` functions, but we surely should make our `find_files` into a generator so we don't have to step through the entire METS for just one match.

Reply:
> There might be many images (even derived images) for that pageId already. It's probably better than nothing though. 

We could retrieve the PAGE XML by pageId and use the `imageFilename` attribute to find the original image by URL - but that will be even more inefficient I guess.

> but we surely should make our find_files into a generator so we don't have to step through the entire METS for just one match.

Good idea, on it.

Reply:
https://github.com/OCR-D/core/issues/613